### PR TITLE
[FLINK-4093] Expose metric interfaces

### PR DESCRIPTION
**Pull request by zentol**
This PR allows users to create their own Counters and register them with Flink.

To that end, Counter was refactored into a SimpleCounter class implementing a new Counter interface. Gauge was changed to an interface as well.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java [hunk 1]
```diff
@@ -146,7 +147,16 @@ public Counter counter(int name) {
 
 	@Override
 	public Counter counter(String name) {
-		Counter counter = new Counter();
+		return counter(name, new SimpleCounter());
+	}
+	
+	@Override
+	public <C extends Counter> C counter(int name, C counter) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java]: What's the advantage of introducing the bounded generic parameter over simply having a `Counter` parameter?

**Comment by tillrohrmann** [comment]
The changes look good to me. +1 for merging. Good work @zentol. I only had a question concerning the use of generic parameters which is not really necessary, imo.

File: flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java (hunk 1 above)
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java]: users can write ```MyCounter c = metricGroup.addCounter("myC", new MyCounter());```

Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java]: in other words, it allows the instantiation, registration and assignment to a variable or field in a single line without loosing the subclass type.

Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java]: Ah true. Good idea :-) Maybe we should change it for the `gauge` and `histogram` functions as well.

Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java]: we should change it for the other methods as well.

**Comment by zentol** [comment]
merging

**Pull request closed without merging**